src/migrate/index.js

```diff
@@ -121,21 +139,72 @@ export default class Migrator {
     });
   }
 
+  _createMigrationLockTable(tableName) {
+    return this.knex.schema.createTable(tableName, function(t) {
+      t.boolean('is_locked');
+    });
+  }
+
+  _getLockTableName(tableName) {
+    return tableName + '_lock';
+  }
+
+  _isLocked() {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .select('*')
+      .then(function(data) {
+        if (data[0].is_locked) {
+          return true;
+        }
+        return false;
+      });
+  }
+
+  _lockMigrations() {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .update({ is_locked: true });
+  }
+
+  _unlockMigrations() {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .update({ is_locked: false });
+  }
+
   // Run a batch of current migrations, in sequence.
   _runBatch(migrations, direction) {
-    return Promise.all(_.map(migrations, this._validateMigrationStructure, this))
+    return this._isLocked()
+      .then(isLocked => {
+        if (isLocked) {
+          throw new Error('migrations failed: migration in progress');
+        }
+        return Promise.all(_.map(migrations, this._validateMigrationStructure, this))
+      })
+      .then(() => this._lockMigrations())
       .then(() => this._latestBatchNumber())
-      .then((batchNo) => {
+      .then(batchNo => {
         if (direction === 'up') batchNo++;
         return batchNo;
       })
-      .then((batchNo) => {
+      .then(batchNo => {
         return this._waterfallBatch(batchNo, migrations, direction)
       })
-      .catch((error) => {
-        helpers.warn('migrations failed with error: ' + error.message)
-        throw error
+      .then(() => this._unlockMigrations())
+      .catch(this.LockError, error => { 
+        helpers.warn('migrations failed with error: ' + error.message);
+        throw error;
       })
+      .catch(error => {
+        helpers.warn('migrations failed with error: ' + error.message)
+        // If the error was not due to a locking issue, then
+        // remove the lock. Otherwise, leave it in place to
+        // force manual intervention
+        return this._unlockMigrations().then(function() {
```

Comment:
use `.finally` instead of then, to make sure that error is thrown also if unlock fails.